Source/WTF/wtf/glib/RunLoopGLib.cpp

```diff
@@ -86,48 +84,65 @@ RunLoop::RunLoop()
 RunLoop::~RunLoop()
 {
     g_source_destroy(m_source.get());
+    m_shouldStop = true;
+}
+
+void RunLoop::runGLibMainLoop(GMainContext* mainContext)
+{
+    g_main_context_ref(mainContext);
+    g_main_context_push_thread_default(mainContext);
+    notifyActivity(Activity::Entry);
+
+    Vector<GPollFD, 32> pollFDs;
+    while (!m_shouldStop) {
+        gint maxPriority = 0;
+        g_main_context_prepare(mainContext, &maxPriority);
+
+        auto numFDs = g_main_context_query(mainContext, maxPriority, nullptr, nullptr, 0);
+        pollFDs.clear();
+        pollFDs.resize(numFDs);
+        auto* pollFDsData = pollFDs.mutableSpan().data();
+
+        gint timeout = 0;
+        g_main_context_query(mainContext, maxPriority, &timeout, pollFDsData, numFDs);
 
-    for (int i = m_mainLoops.size() - 1; i >= 0; --i) {
-        if (!g_main_loop_is_running(m_mainLoops[i].get()))
-            continue;
-        g_main_loop_quit(m_mainLoops[i].get());
+        notifyActivity(Activity::BeforeWaiting);
```

Comment:
[Documentation says:](https://developer.apple.com/documentation/corefoundation/cfrunloopactivity/beforewaiting)

> Inside the event processing loop before the run loop sleeps, waiting for a source or timer to fire. This activity does not occur if [CFRunLoopRunInMode(_:_:_:)](https://developer.apple.com/documentation/corefoundation/cfrunloopruninmode(_:_:_:)) is called with a timeout of 0 seconds. It also does not occur in a particular iteration of the event processing loop if a version 0 source fires.

I think that means we should notify it only if `timeout > 0`.

Reply:
Still need to tackle this.

Reply:
I'm not sure about this anymore:

```
This activity does not occur if [CFRunLoopRunInMode(_:_:_:)](https://developer.apple.com/documentation/corefoundation/cfrunloopruninmode(_:_:_:)) is called with a timeout of 0 seconds.
```

it specifically talks about `CFRunLoopRunInMode` called with a timeout of 0.


The CF RunLoop implementation never specifies a timeout of 0. It either calls `CFRunLoopRunInMode` with a `0.05` timeout in `cycle()`, or the `CFRunLoopRun` variant that spins the loop indefinately.

```
RunLoop::CycleResult RunLoop::cycle(RunLoopMode mode)
{
    CFTimeInterval timeInterval = 0.05;
    CFRunLoopRunInMode(mode, timeInterval, true);
    return CycleResult::Continue;
}

void RunLoop::run()
{
    AutodrainedPool pool;
    CFRunLoopRun();
}
```